File: SciPy/cminpack-master/examples/hybdrv_.c
/* Usage: lmddrv < lm.data */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "minpack.h"
#include "vec.h"
#define real __minpack_real__

/*     ********** */

/*     this program tests codes for the solution of n nonlinear */
/*     equations in n variables. it consists of a driver and an */
/*     interface subroutine fcn. the driver reads in data, calls the */
/*     nonlinear equation solver, and finally prints out information */
/*     on the performance of the solver. this is only a sample driver, */
/*     many other drivers are possible. the interface subroutine fcn */
/*     is necessary to take into account the forms of calling */
/*     sequences used by the function subroutines in the various */
/*     nonlinear equation solvers. */

/*     subprograms called */

/*       user-supplied ...... fcn */

/*       minpack-supplied ... dpmpar,enorm,hybrd1,initpt,vecfcn */

/*       fortran-supplied ... dsqrt */

/*     argonne national laboratory. minpack project. march 1980. */
/*     burton s. garbow, kenneth e. hillstrom, jorge j. more */

/*     ********** */

void fcn(const int *n, const real *x, real *fvec, int *iflag);

struct refnum {
    int nprob, nfev, njev;
};

struct refnum hybrdtest;

static void printvec(int n, const real *x)
{
    int i, num5, ilow, numleft;
    num5 = n/5;

    for (i = 0; i < num5; ++i) {
        ilow = i*5;
        printf("     %15.7e%15.7e%15.7e%15.7e%15.7e\n",
               (double)x[ilow+0], (double)x[ilow+1], (double)x[ilow+2], (double)x[ilow+3], (double)x[ilow+4]);
    }
    
    numleft = n%5;
    ilow = n - numleft;

    switch (numleft) {
        case 1:
            printf("     %15.7e\n",
                   (double)x[ilow+0]);
            break;
        case 2:
            printf("     %15.7e%15.7e\n",
                   (double)x[ilow+0], (double)x[ilow+1]);
            break;
        case 3:
            printf("     %15.7e%15.7e%15.7e\n",
                   (double)x[ilow+0], (double)x[ilow+1], (double)x[ilow+2]);
            break;
        case 4:
            printf("     %15.7e%15.7e%15.7e%15.7e\n",
                   (double)x[ilow+0], (double)x[ilow+1], (double)x[ilow+2], (double)x[ilow+3]);
            break;
    }
}

/* Main program */
int main(int argc, char **argv)
{

    int i,ic,k,n,ntries;
    int info;

    int na[60];
    int nf[60];
    int nj[60];
    int np[60];
    int nx[60];

    real factor,fnorm1,fnorm2,tol;

    real fnm[60];
    real fvec[40];
    real x[40];

    real wa[2660];
    const int lwa = 2660;
    const int i1 = 1;
    (void)argc; (void)argv;

    tol = sqrt(__minpack_func__(dpmpar)(&i1));

    ic = 0;

    for (;;) {
        scanf("%5d%5d%5d\n", &hybrdtest.nprob, &n, &ntries);
        if (hybrdtest.nprob <= 0.)
            break;

        factor = 1.;

        for (k = 0; k < ntries; ++k, ++ic) {
            hybipt(n,x,hybrdtest.nprob,factor);

            vecfcn(n,x,fvec,hybrdtest.nprob);

            fnorm1 = __minpack_func__(enorm)(&n,fvec);

            printf("\n\n\n\n      problem%5d      dimension%5d\n\n", hybrdtest.nprob, n);

            hybrdtest.nfev = 0;
            hybrdtest.njev = 0;

            __minpack_func__(hybrd1)(fcn,&n,x,fvec,&tol,&info,wa,&lwa);

            fnorm2 = __minpack_func__(enorm)(&n,fvec);

            np[ic] = hybrdtest.nprob;
            na[ic] = n;
            nf[ic] = hybrdtest.nfev;
            hybrdtest.njev /= n;
            nj[ic] = hybrdtest.njev;
            nx[ic] = info;

            fnm[ic] = fnorm2;

            printf("\n      initial l2 norm of the residuals%15.7e\n"
                   "\n      final l2 norm of the residuals  %15.7e\n"
                   "\n      number of function evaluations  %10d\n"
                   "\n      number of Jacobian evaluations  %10d\n"
                   "\n      exit parameter                  %10d\n"
                   "\n      final approximate solution\n\n",
                   (double)fnorm1, (double)fnorm2, hybrdtest.nfev, hybrdtest.njev, info);
            printvec(n, x);

            factor *= 10.;

        }

    }

    printf("\f summary of %d calls to hybrd1\n", ic);
    printf("\n nprob   n    nfev   njev  info  final l2 norm\n\n");
    for (i = 0; i < ic; ++i) {
        printf("%4d%6d%7d%7d%6d%16.7e\n",
               np[i], na[i], nf[i], nj[i], nx[i], (double)fnm[i]);
    }
    exit(0);
}


void fcn(const int *n, const real *x, real *fvec, int *iflag)
{
/*     ********** */

/*     the calling sequence of fcn should be identical to the */
/*     calling sequence of the function subroutine in the nonlinear */
/*     equation solver. fcn should only call the testing function */
/*     subroutine vecfcn with the appropriate value of problem */
/*     number (nprob). */

/*     subprograms called */

/*       minpack-supplied ... vecfcn */

/*     argonne national laboratory. minpack project. march 1980. */
/*     burton s. garbow, kenneth e. hillstrom, jorge j. more */

/*     ********** */
    vecfcn(*n, x, fvec, hybrdtest.nprob);
    if (*iflag == 1) {
        hybrdtest.nfev++;
    }
    if (*iflag == 2) {
        hybrdtest.njev++;
    }
} /* fcn_ */

File: SciPy/cminpack-master/examples/hybdrv.f
c     **********
c
c     this program tests codes for the solution of n nonlinear
c     equations in n variables. it consists of a driver and an
c     interface subroutine fcn. the driver reads in data, calls the
c     nonlinear equation solver, and finally prints out information
c     on the performance of the solver. this is only a sample driver,
c     many other drivers are possible. the interface subroutine fcn
c     is necessary to take into account the forms of calling
c     sequences used by the function subroutines in the various
c     nonlinear equation solvers.
c
c     subprograms called
c
c       user-supplied ...... fcn
c
c       minpack-supplied ... dpmpar,enorm,hybrd1,initpt,vecfcn
c
c       fortran-supplied ... dsqrt
c
c     argonne national laboratory. minpack project. march 1980.
c     burton s. garbow, kenneth e. hillstrom, jorge j. more
c
c     **********
      integer i,ic,info,k,lwa,n,nfev,nprob,nread,ntries,nwrite
      integer na(60),nf(60),np(60),nx(60)
      double precision factor,fnorm1,fnorm2,one,ten,tol
      double precision fnm(60),fvec(40),wa(2660),x(40)
      double precision dpmpar,enorm
      external fcn
      common /refnum/ nprob,nfev
c
c     logical input unit is assumed to be number 5.
c     logical output unit is assumed to be number 6.
c
      data nread,nwrite /5,6/
c
      data one,ten /1.0d0,1.0d1/
      tol = dsqrt(dpmpar(1))
      lwa = 2660
      ic = 0
   10 continue
         read (nread,50) nprob,n,ntries
         if (nprob .le. 0) go to 30
         factor = one
         do 20 k = 1, ntries
            ic = ic + 1
            call initpt(n,x,nprob,factor)
            call vecfcn(n,x,fvec,nprob)
            fnorm1 = enorm(n,fvec)
            write (nwrite,60) nprob,n
            nfev = 0
            call hybrd1(fcn,n,x,fvec,tol,info,wa,lwa)
            fnorm2 = enorm(n,fvec)
            np(ic) = nprob
            na(ic) = n
            nf(ic) = nfev
            nx(ic) = info
            fnm(ic) = fnorm2
            write (nwrite,70) fnorm1,fnorm2,nfev,info,(x(i), i = 1, n)
            factor = ten*factor
   20       continue
         go to 10
   30 continue
      write (nwrite,80) ic
      write (nwrite,90)
      do 40 i = 1, ic
         write (nwrite,100) np(i),na(i),nf(i),nx(i),fnm(i)
   40    continue
      stop
   50 format (3i5)
   60 format ( //// 5x, 8h problem, i5, 5x, 10h dimension, i5, 5x //)
   70 format (5x, 33h initial l2 norm of the residuals, d15.7 // 5x,
     *        33h final l2 norm of the residuals  , d15.7 // 5x,
     *        33h number of function evaluations  , i10 // 5x,
     *        15h exit parameter, 18x, i10 // 5x,
     *        27h final approximate solution // (5x, 5d15.7))
   80 format (12h1summary of , i3, 16h calls to hybrd1 /)
   90 format (39h nprob   n    nfev  info  final l2 norm /)
  100 format (i4, i6, i7, i6, 1x, d15.7)
c
c     last card of driver.
c
      end
      subroutine fcn(n,x,fvec,iflag)
      integer n,iflag
      double precision x(n),fvec(n)
c     **********
c
c     the calling sequence of fcn should be identical to the
c     calling sequence of the function subroutine in the nonlinear
c     equation solver. fcn should only call the testing function
c     subroutine vecfcn with the appropriate value of problem
c     number (nprob).
c
c     subprograms called
c
c       minpack-supplied ... vecfcn
c
c     argonne national laboratory. minpack project. march 1980.
c     burton s. garbow, kenneth e. hillstrom, jorge j. more
c
c     **********
      integer nprob,nfev
      common /refnum/ nprob,nfev
      call vecfcn(n,x,fvec,nprob)
      nfev = nfev + 1
      return
c
c     last card of interface subroutine fcn.
c
      end
